File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <vector>

using namespace std;

static string cmds[4] = {"d", "l", "r", "u"};
static int dr[4] = {1, 0, 0, -1};
static int dc[4] = {0, -1, 1, 0};

static string result = "z";
static int N, M, R, C, K;

void dfs(int, int, int, string);
int calc_distance(int, int, int, int);
bool isImpossible(int, int);

string solution(int n, int m, int x, int y, int r, int c, int k) {
    // n, m: 격자 크기
    // x, y: 출발 위치
    // r, c: 도착 위치
    // k: 이동 횟수
    N = n; M = m; R = r; C = c; K = k;
    
    if (isImpossible(x, y))
        return "impossible";
    
    string command = "";
    dfs(x, y, 0, command);
    
    string answer = result;
    return answer;
}

void dfs(int x, int y, int moves, string command) {
    
    // 현재 이동횟수 + 목적지까지의 최단거리 > k면 종료. 길을 잘못 든 것.
    if ((moves + calc_distance(x, y, R, C)) > K)
        return;
    
    if (command > result)
        return;
    
    if (moves == K && x == R && y == C) {
        result = (result < command) ? result : command;
        return;
    }
    
    int nr, nc;
    for(int i=0; i<4; i++) {
        nr = x + dr[i];
        nc = y + dc[i];
        
        if (nr > 0 && nc > 0 && nr <= N && nc <= M) {
            dfs(nr, nc, moves + 1, command + cmds[i]);
        }
        
    }
    
}

int calc_distance(int x, int y, int r, int c) {
    return std::abs(x-r) + std::abs(y-c);
}

// 목적지까지 최단거리가 k보다 크거나, (k-최단거리)가 짝수가 아니라면 이동 불가
bool isImpossible(int x, int y) {
    int d = calc_distance(x, y, R, C);
    
    if (d > K || (K-d)%2 != 0)
        return true;
    
    return false;
}

File: 프로그래머스/3/150365. 미로 탈출 명령어/미로 탈출 명령어.java
import java.util.*;
import java.io.*;
import java.math.*;

class Solution {
    
    static int[] dr = {1, 0, 0, -1};
    static int[] dc = {0, -1, 1, 0};
    static String[] cmds = {"d", "l", "r", "u"};
    
    static String result = "z";
    static int N, M, K;
    
    public String solution(int n, int m, int x, int y, int r, int c, int k) {
        // n, m은 격자 크기
        // x, y는 출발지 좌표
        // r, c는 목적지 좌표
        // k는 이동 횟수
        N = n; M = m; K=k;
        
        String answer = "";
        
        if (isImpossible(x, y, r, c))
            answer = "impossible";
        else {
            dfs(r, c, x, y, 0, "");
            // System.out.println("result = " + result);
            answer = result;
        }
        
        // System.out.println(answer);
        return answer;
    }
    
    static void dfs(int e_r, int e_c, int r, int c, int moves, String cmd) {
        
        // moves + 현재 위치에서 목적지까지 최단거리가 k보다 크면 리턴
        // if (isFailRoute(r, c, e_r, e_c, moves))
        if (isFailRoute(r, c, e_r, e_c, moves) || cmd.compareTo(result) > 0)
            return;
        
        // 제일 먼저 찾은 정답이 가장 빠른 명령어 조합임. cmds 추가하는 순서가 그렇게 되어있음.
        if (moves == K && r == e_r && c == e_c) {
            // result = (result.compareTo(cmd) > 0) ? cmd : result;
            result = cmd;
            return;
        }
        
        int nr, nc;
        for(int i=0; i<4; i++) {
            nc = c + dc[i];
            nr = r + dr[i];
            if (nr > 0 && nc > 0 && nr <= N && nc <= M) {
                dfs(e_r, e_c, nr, nc, moves+1, cmd + cmds[i]);
            }
        }
        
    }
    
    static boolean isImpossible(int r, int c, int e_r, int e_c) {
        int lr = e_r - r;
        int lc = e_c - c;
        
        int distance = Math.abs(lr) + Math.abs(lc);
        if (distance > K || (K - distance) % 2 != 0)
            return true;
        else
            return false;
    }
    
    static boolean isFailRoute(int r, int c, int e_r, int e_c, int moves) {
        int lr = e_r - r;
        int lc = e_c - c;
        
        int distance = Math.abs(lr) + Math.abs(lc);
        return (moves + distance) > K;
    }
    
}